sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java

```diff
@@ -0,0 +1,662 @@
+    private final String tempGcsPath;
+    private final String deadLetterGcsPath;
+    private final ContentStructure contentStructure;
+
+    Import(
+        String fhirStore,
+        String tempGcsPath,
+        String deadLetterGcsPath,
+        @Nullable ContentStructure contentStructure) {
+      this.fhirStore = fhirStore;
+      this.tempGcsPath = tempGcsPath;
+      this.deadLetterGcsPath = deadLetterGcsPath;
+      if (contentStructure == null) {
+        this.contentStructure = ContentStructure.CONTENT_STRUCTURE_UNSPECIFIED;
+      } else {
+        this.contentStructure = contentStructure;
+      }
+    }
+
+    @Override
+    public PCollection<HealthcareIOError<HttpBody>> expand(PCollection<HttpBody> input) {
+      return input.apply(
+          ParDo.of(new ImportFn(fhirStore, tempGcsPath, deadLetterGcsPath, contentStructure)));
+    }
+
+    public enum ContentStructure {
+      /** If the content structure is not specified, the default value BUNDLE will be used. */
+      CONTENT_STRUCTURE_UNSPECIFIED,
+      /**
+       * The source file contains one or more lines of newline-delimited JSON (ndjson). Each line is
+       * a bundle, which contains one or more resources. Set the bundle type to history to import
+       * resource versions.
+       */
+      BUNDLE,
+      /**
+       * The source file contains one or more lines of newline-delimited JSON (ndjson). Each line is
+       * a single resource.
+       */
+      RESOURCE,
+      /** The entire file is one JSON bundle. The JSON can span multiple lines. */
+      BUNDLE_PRETTY,
+      /** The entire file is one JSON resource. The JSON can span multiple lines. */
+      RESOURCE_PRETTY
+    }
+
+    static class ImportFn extends DoFn<HttpBody, HealthcareIOError<HttpBody>> {
+      private final String fhirStore;
+      private final String tempGcsPath;
+      private final String deadLetterGcsPath;
+      private final ContentStructure contentStructure;
+      private ObjectMapper mapper;
+      private ResourceId resourceId;
+      private ResourceId deadLetterResourceId;
+      private WritableByteChannel ndJsonChannel;
+
+      private transient HealthcareApiClient client;
+      private static final Logger LOG = LoggerFactory.getLogger(ImportFn.class);
+
+      ImportFn(
+          String fhirStore,
+          String tempGcsPath,
+          String deadLetterGcsPath,
+          @Nullable ContentStructure contentStructure) {
+        this.fhirStore = fhirStore;
+        this.tempGcsPath = tempGcsPath;
+        this.deadLetterGcsPath = deadLetterGcsPath;
+        if (contentStructure == null) {
+          this.contentStructure = ContentStructure.CONTENT_STRUCTURE_UNSPECIFIED;
+        } else {
+          this.contentStructure = contentStructure;
+        }
+      }
+
+      @Setup
+      public void initClient() throws IOException {
+        this.client = new HttpHealthcareApiClient();
+      }
+
+      @StartBundle
+      public void initBatch() throws IOException {
```

Comment:
@lastomato @yeweidaniel @ygupta89 @pabloem 
Please review the logic here and LMK if you feel this is spamming the import API.
As written this will create a separate new line delimited JSON file per bundle (group of elements in a `PCollection` of arbitrary size decided by runner) and import it w/ a separate call to import API.

This was the "easiest" way to implement but may not be the most appropriate usage of the import API.

PTAL and LMK if you'd like this to refactored into something that:
1) asserts bounded PCollection
2) waits til all temp files are written 
3) blocks on import by polling the operation til complete.

It's not immediately obvious to me how to achieve 1 & 2 in the beam framework but I can do some research.

Reply:
I am a little worried about the fact that the bundle size is unknown. In extreme (though unlikely) cases, we could be making one import API call per resource.

Reply:
Research:

- Based on [this SO post](https://stackoverflow.com/questions/44255924/error-message-too-many-sources-provided-15285-limit-is-10000) 
We might be able to `GroupByKey` single dummy key to encourage runner to do less bundle splitting, though I'm not sure that will really do the trick. 
- From Beam docs:
> Specifically, it makes it impossible to batch any operations, such as writing elements to a sink or checkpointing progress during processing.
>The division of the collection into bundles is arbitrary and selected by the runner. This allows the runner to choose an appropriate middle-ground between persisting results after every element, and having to retry everything if there is a failure. For example, a streaming runner may prefer to process and commit small bundles, and a batch runner may prefer to process larger bundles.
([source](https://beam.apache.org/documentation/runtime/model/#bundling-and-persistence))

It seems to me that our predominant use case for Import is a batch pipeline and this sentence implies that bundles are larger in batch (not sure if we can find internal details on Dataflow Runner). 

- BigQueryIO has a [batch loads](https://github.com/apache/beam/blob/master/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java) pattern which seems similar to what we are aiming to achieve here. This seems quite complex and I'm wondering if there are simpler approaches elsewhere in the beam code base.

Do we have any guidance on how many parallel import operations would be acceptable? Perhaps we can run some benchmarks with 500, 1000, 10000 resources with this approach / adding group by single key and see if it's acceptable.

Reply:
@pabloem is there a good way to signal to the runner to prefer large bundles for this DoFn? Are there any other common patterns for batching files for bulk import operations that are less complicated than the BQ File Loads (which seem to have a lot of complexity to handle dynamic destinations)?